File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/1-Karatsuba Multiplication/Karatsuba.java
import java.math.BigInteger;

public class Karatsuba {

    /**
     * Calculates the multiplication of two integers using Karatsuba
     * Multiplication.
     * Naive calculation: O(n^2)
     * @param x first operand
     * @param y second operand
     * @return product
     */
    private BigInteger karatsuba(BigInteger x, BigInteger y) {
        // We assume the input x and y are both non-negative.

        String xStr = x.toString(), yStr = y.toString();

        // Pad leading zeros to make x and y of the same length   [O(n)]
        if (xStr.length() > yStr.length()) {
            yStr = padZeros(yStr, xStr.length() - yStr.length(), true);
        } else if (xStr.length() < yStr.length()) {
            xStr = padZeros(xStr, yStr.length() - xStr.length(), true);
        }

        int n = xStr.length();
        // Base case
        if (n == 1) {
            return x.multiply(y);
        }
        // Recursive case
        // [Divide]   [O(n)]
        String aStr = xStr.substring(0, n / 2), bStr = xStr.substring(n / 2), cStr = yStr.substring(0, n / 2),
                dStr = yStr.substring(n / 2);
        BigInteger a = new BigInteger(aStr), b = new BigInteger(bStr), c = new BigInteger(cStr),
                d = new BigInteger(dStr);
        // [Conquer]
        BigInteger ac = karatsuba(new BigInteger(aStr), c);
        BigInteger bd = karatsuba(b, d);
        BigInteger adBC = karatsuba(a.add(b), c.add(d)).subtract(ac).subtract(bd);
        // Combine the results   [O(n)]
        String part1Str = padZeros(ac.toString(), 2 * (n - n / 2), false);
        String part2Str = padZeros(adBC.toString(), n - n / 2, false);
        return new BigInteger(part1Str).add(new BigInteger(part2Str)).add(bd);
        // T(n) = 3T(n/2) + O(n)
        // a = 2, b = 2, d = 1
        // According to Master Method, the overall running time complexity is O(n^1.585), better than O(n^2).
    }

    /**
     * Private helper method to pad the given number of zeros to the given
     * string.
     * @param s string to be padded
     * @param nZero number of zeros
     * @param atFront boolean whether to pad zeros at front
     * @return string with padded zeros
     */
    private String padZeros(String s, int nZero, boolean atFront) {
        StringBuilder newStr = new StringBuilder();
        if (atFront) {
            for (int i = 0; i < nZero; ++i) {
                newStr.append(0);
            }
            newStr.append(s);
        } else {
            newStr.append(s);
            for (int i = 0; i < nZero; ++i) {
                newStr.append(0);
            }
        }
        return newStr.toString();
        // Running time complexity: O(n)
    }

}

File: 1-Divide and Conquer, Sorting and Searching, and Randomized Algorithms/1-Karatsuba Multiplication/karatsuba.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

__author__ = 'Ziang Lu'


def karatsuba(x: int, y: int) -> int:
    """
    Calculates the multiplication of two integers using Karatsuba
    Multiplication.
    Naive calculation: O(n^2)
    :param x: int
    :param y: int
    :return: int
    """
    # We assume that the input x and y are both non-negative.

    x_s, y_s = str(x), str(y)

    # Pad leading zeros   [O(n)]
    if len(x_s) > len(y_s):
        y_s = _pad_zeros(y_s, len(x_s) - len(y_s), at_front=True)
    elif len(x_s) < len(y_s):
        x_s = _pad_zeros(x_s, len(y_s) - len(x_s), at_front=True)

    n = len(x_s)
    # Base case
    if n == 1:
        return x * y
    # Recursive case
    # [Divide]   [O(n)]
    a_s, b_s, c_s, d_s = x_s[:n // 2], x_s[n // 2:], y_s[:n // 2], y_s[n // 2:]
    a, b, c, d = int(a_s), int(b_s), int(c_s), int(d_s)
    # [Conquer]
    ac = karatsuba(a, c)
    bd = karatsuba(b, d)
    ad_bc = karatsuba(a + b, c + d) - ac - bd
    # Combine the results   [O(n)]
    part_1_s = _pad_zeros(str(ac), 2 * (n - n // 2), at_front=False)
    part_2_s = _pad_zeros(str(ad_bc), n - n // 2, at_front=False)
    return int(part_1_s) + int(part_2_s) + bd
    # T(n) = 3T(n/2) + O(n)
    # a = 2, b = 2, d = 1
    # According to Master Method, the overall running time complexity is
    # O(n^1.585), better than O(n^2).


def _pad_zeros(s: str, n_zeros: int, at_front: bool) -> str:
    """
    Private helper function to pad the given number of zeros to the given
    string.
    :param s: str
    :param n_zeros: int
    :param at_front: bool
    :return: str
    """
    new_s = ""
    if at_front:
        new_s += '0' * n_zeros
        new_s += s
    else:
        new_s += s
        new_s += '0' * n_zeros
    return new_s
    # Running time complexity: O(n)
